### PR TITLE
Add Unity3D grammar

### PR DESCRIPTION
**Pull request by TTeig**
# Unity3D grammar

## Description

I have added grammar for unity. They should be pretty self-explanatory, but I'd like other unity users to test it and give input on naming.

## Motivation and Context

Several users showed interest in a Unity grammar.

## How Has This Been Tested
Manually testing all the commands within unity. 

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue or bug)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Renamed existing command phrases (we discourage this without a strong rationale).

## Checklist

- [x] I have read the CONTRIBUTING document.
- [x] My code follows the code style of this project.
- [x] I have checked that my code does not duplicate functionality elsewhere in Caster.
- [x] I have checked for and utilized existing command phrases from within Caster (delete if not applicable). 
- [x] My code implements all the features I wish to merge in this pull request.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests pass.

## Maintainer/Reviewer Checklist

- [x] Basic functionality has been tested and works as claimed.
- [x] New documentation is clear and complete.
- [x] Code is clear and readable.

### CODE REVIEW

**Comment by TTeig** [comment]
I'll hold off on updating the documentation until I've had some eyes on it. We can always save the merge until then.

**Comment by kendonB** [comment]
@TTeig I'm not sure we have any other Unity users in the community (you could ask on gitter). Perhaps ask a colleague to test it out?

**Comment by TTeig** [comment]
@qubitz, do you have time to have a look? There's one more person on gitter I couldn't ping here for some reason, but I'll message him. I made the PR since they were both interested in a unity grammar.

**Comment by TTeig** [comment]
if they are not around I can definitely find someone else to look so we can close this one off.

**Comment by davidlehub** [comment]
I use Unity. And has try this Unity grammar, so far so good :)

**Comment by davidlehub** [comment]
what is "PR" means again?

**Comment by TTeig** [comment]
pull request.

**Comment by davidlehub** [comment]
For sure, be able to control Unity by voice + the mouse, will be great. But since I'm learning Unity, I don't use voice command a lot yet. But, later it gonna be very usefully.

**Comment by davidlehub** [comment]
> pull request.

hmm, i just copy the Unity grammar .py file and put it in Caster folder. Because i don't know much about github, and don't wanna make mistakes :D

**Comment by davidlehub** [comment]
But where to pull request btw??

**Comment by TTeig** [comment]
I can message you and walk you through it best I can.

**Comment by davidlehub** [comment]
i guess this button?:
image

File: castervoice/rules/apps/editor/unity.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from dragonfly import Key,MappingRule, Dictation, Choice,Text
+from castervoice.lib.merge.state.short import R
+from castervoice.lib.merge.additions import IntegerRef
+from castervoice.lib.ctrl.mgr.rule_details import RuleDetails
+
+class UnityRule(MappingRule):
+	mapping = {
+		"<window>" : R(Key("c-%(window)s")),
```
Comment (qubitz) [inline castervoice/rules/apps/editor/unity.py:8]: Personally, I prefer keywords before commands like this to reduce vocabulary clutter/conflict. The first thing that comes to mind is "go <window>". Is this supposed to open up new windows as well?

File: castervoice/rules/apps/editor/unity.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		"search <searchable> [<dictation>]" : R(Key("c-%(searchable)s,c-f") + Text("%(dictation)s")),
```
Comment (qubitz) [inline castervoice/rules/apps/editor/unity.py:9]: Wow, had no idea a command like this existed!

File: castervoice/rules/apps/editor/unity.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		"pause game" : R(Key("cs-p")),
+
+		"build and run" : R(Key("c-b")),
+		"build settings" : R(Key("cs-b")),
+
+		"make empty" : R(Key("as-n")),
+		"add component" : R(Key("cs-a")),
+
+		"align with view" : R(Key("cs-f")),
+		"move to view" : R(Key("ca-f")),
```
Comment (qubitz) [inline castervoice/rules/apps/editor/unity.py:25]: Super settle, but how do you feel about renaming these to "view align" and "view move"? I'm kind of torn myself on the question.

**Comment by qubitz** [comment]
Thanks @TTaig for adding me. I don't browse the pull requests enough and would've totally missed this. I left a couple of comments, but overall I'm looking forward to using these! What's the best way to suggest additional commands to this grammar? I'd love to add "reload [scripts]" and "rename".

File: castervoice/rules/apps/editor/unity.py (hunk 1 above)
Comment (TTeig) [reply castervoice/rules/apps/editor/unity.py:8]: Excellent point. Other suggestions could be "show" or "focus". I will think about what feels best.

(Triggering the shortcut for a window in Unity will shift focus to it or open it if it's not already part of the active layout.)

File: castervoice/rules/apps/editor/unity.py (hunk 3 above)
Comment (TTeig) [reply castervoice/rules/apps/editor/unity.py:25]: Personally I don't mind slightly more verbose commands for things that are not used too often if it makes them more memorable, descriptive and consistent. The commands are the same as the description you find in the unity menus. Saying just "view move" makes it sound like you're moving the view itself and would conflict with other commands for interacting with views whatever that would be.

If anything I think the commands I picked should be the official commands in this case, but we could add a shorter alias. Although that is also something individual users can do themselves.

File: castervoice/rules/apps/editor/unity.py (hunk 2 above)
Comment (TTeig) [reply castervoice/rules/apps/editor/unity.py:9]: It doesn't in Unity. It's just a series of key strokes compressed to one command. For Caster commands in general I love the ones that let me add dictation without a break. And there's a lot of potential for creating more commands in Unity for any actions that take the same input every time.

**Comment by TTeig** [comment]
@qubitz  You can make make the pull request yourself, or add a new feature request issue. Tag me and I might get around to do it when I have time or ping me on Gitter. (I only just started contributing myself so I'm no authority, but you can always ask on Gitter about best practices.)

F2 is a pretty standard shortcut for renaming so it's not strictly needed, but adding a `rename [to] <dictation>` command is on top of my list. But remind me how you reload scripts in Unity.

**Comment by qubitz** [comment]
I turned off Unity's auto-reloading option to prevent unwanted errors from appearing too soon so it's unnecessary for most users, but the command is Ctrl+R.

**Comment by qubitz** [comment]
Thanks for the changes! 👍. Although I did notice that [nav2.py](https://github.com/dictation-toolbox/Caster/blob/master/castervoice/rules/core/navigation_rules/nav2.py) already has `refresh`, my bad.

**Comment by TTeig** [comment]
@qubitz Sorted.

**Comment by TTeig** [comment]
Well, that's strange. Not sure why Travis would fail and I can't tell from the log.

**Comment by LexiconCode** [comment]
Is this considered complete?

**Comment by TTeig** [comment]
Yes, I'm happy with it.

**Pull request merged**